### PR TITLE
[SPARK-28770][CORE][TESTS]Ignore SparkListenerStageExecutorMetrics in testApplicationReplay test

### PR DESCRIPTION
**Pull request by bzhaoopenstack**
### What changes were proposed in this pull request?
This patch fix the testApplicationReplay test in ReplayListenerSuite.
We compare the event log json except SparkListenerStageExecutorMetrics.

### Why are the changes needed?
As for potentail different log events Issue, we found this can be reproduced in X86. This due to the SparkListenerStageExecutorMetrics events may be logged in different order. So the test would be fail in some time. But this is a particular issue for Arm platform. For more details, please see
https://issues.apache.org/jira/browse/SPARK-28770

### Does this PR introduce any user-facing change?
No

### How was this patch tested?
This is existing UTs, we can test it with the same way like we usually do.

### CODE REVIEW

**Comment by bzhaoopenstack** [comment]
@wypoon @srowen 
Hi guys, cloud you please have a look at this patch? Thank you very much

**Review by srowen — COMMENTED** [review]
Instead of ignoring them, is there any natural sort ordering you can impose to make the comparison valid? does the test fail right now?

**Comment by wypoon** [comment]
`SparkListenerStageExecutorMetrics` were introduced in SPARK-23429 (https://github.com/apache/spark/pull/21221/). By design, executor metrics update events are not logged by the `EventLoggingListener`. Instead, the listener keeps track of the per stage peaks for any of the executors and the driver for which it has received metrics. On stage completion, the peaks for the stage are logged in `SparkListenerStageExecutorMetrics` events for each of these executors and driver.
Since executor metrics update events are not logged in the event log, they do not get replayed. Thus the listener for the replay never sees metrics updates. It is therefore valid to exclude `SparkListenerStageExecutorMetrics` events from both the original and the replay for the purpose of comparison.

However, instead of excluding all `SparkListenerStageExecutorMetrics` events from both the original `EventLoggingListener` and the replay listener, we can have a finer-grained fix, which I have proposed in https://github.com/apache/spark/pull/25673/ for comparison. It should be sufficient to exclude any `SparkListenerStageExecutorMetrics` events for the driver. This is because with SPARK-26329 (https://github.com/apache/spark/pull/23767/), executor metrics are also sent in task end events (which do get replayed), so the `EventLoggingListener` always receives metrics for the executors (just not necessarily for the driver), and thus `SparkListenerStageExecutorMetrics` events for the executors always get logged.

File: core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala [hunk 1]
```diff
@@ -218,8 +219,17 @@ class ReplayListenerSuite extends SparkFunSuite with BeforeAndAfter with LocalSp
...
     // Verify the same events are replayed in the same order
     assert(sc.eventLogger.isDefined)
-    val originalEvents = sc.eventLogger.get.loggedEvents
-    val replayedEvents = eventMonster.loggedEvents
+    implicit val format = DefaultFormats
+    def exceptCase(a: JValue) = ( a \ "Event").extract[String] match {
+      // If we are logging stage executor metrics, there is a bulk call to logEvent with
+      // SparkListenerStageExecutorMetrics events via a Map.foreach. The Map.foreach bulk
+      // operation may not log the events with the same order. So here we should not compare
+      // SparkListenerStageExecutorMetrics here.
```
Comment (wypoon) [inline core/src/test/scala/org/apache/spark/scheduler/ReplayListenerSuite.scala]: This comment is based on my earlier analysis in the JIRA, which is not quite correct.

**Comment by HeartSaVioR** [comment]
#25673 supersedes this and addresses root issue. Could you take a look at #25673 and test with ARM CI, and close this PR if #25673 works? Thanks in advance!

**Comment by bzhaoopenstack** [comment]
Thanks very much, guys. #25673 works well on ARM. 
Also thanks to @wypoon, you give us a very clear failure reason we face. Thanks

**Comment by wypoon** [comment]
@bzhaoopenstack, thanks for the verification.

**Pull request closed without merging**